fuse_core/src/async_motion_model.cpp

```diff
@@ -62,56 +62,79 @@ bool AsyncMotionModel::apply(Transaction& transaction)
   // This function blocks until the queryCallback() call completes, thus enforcing that motion models are generated
   // in order.
-  auto callback = boost::make_shared<CallbackWrapper<bool>>(
+  auto callback = std::make_shared<CallbackWrapper<bool>>(
     std::bind(&AsyncMotionModel::applyCallback, this, std::ref(transaction)));
   auto result = callback->getFuture();
-  callback_queue_.addCallback(callback, reinterpret_cast<uint64_t>(this));
+  callback_queue_->addCallback(callback);
   result.wait();
-  return result.get();
-}
 
-void AsyncMotionModel::graphCallback(Graph::ConstSharedPtr graph)
-{
-  callback_queue_.addCallback(
-    boost::make_shared<CallbackWrapper<void>>(std::bind(&AsyncMotionModel::onGraphUpdate, this, std::move(graph))),
-    reinterpret_cast<uint64_t>(this));
+  return result.get();
 }
 
 void AsyncMotionModel::initialize(const std::string& name)
 {
   // Initialize internal state
   name_ = name;
-  node_handle_.setCallbackQueue(&callback_queue_);
-  private_node_handle_ = ros::NodeHandle("~/" + name_);
-  private_node_handle_.setCallbackQueue(&callback_queue_);
+  std::string node_namespace = "";
+
+  rclcpp::Context::SharedPtr ros_context = rclcpp::contexts::get_global_default_context();
+  auto node_options = rclcpp::NodeOptions();
+
+  ros_context->init(0, NULL);    // XXX should expose the init arg list
+  node_options.context(ros_context); //set a context to generate the node in
+
+  node_ = rclcpp::Node::make_shared(name_, node_namespace, node_options);
+
+  auto executor_options = rclcpp::ExecutorOptions();
+  executor_options.context = ros_context;
+  executor_ = rclcpp::executors::MultiThreadedExecutor::make_shared(executor_options, executor_thread_count_);
+
+  callback_queue_ = std::make_shared<CallbackAdapter>(ros_context);
+  node_->get_node_waitables_interface()->add_waitable(
+    callback_queue_, (rclcpp::CallbackGroup::SharedPtr) nullptr);
 
   // Call the derived onInit() function to perform implementation-specific initialization
   onInit();
 
-  // Start the async spinner to service the local callback queue
-  spinner_.start();
+  executor_->add_node(node_);
+}
+
+void AsyncMotionModel::graphCallback(Graph::ConstSharedPtr graph)
+{
+  auto callback = std::make_shared<CallbackWrapper<void>>(
+    std::bind(&AsyncMotionModel::onGraphUpdate, this, std::move(graph))
+  );
+  auto result = callback->getFuture();
+  callback_queue_->addCallback(callback);
+  result.wait();
 }
 
 void AsyncMotionModel::start()
 {
-  auto callback = boost::make_shared<CallbackWrapper<void>>(std::bind(&AsyncMotionModel::onStart, this));
+  auto callback = std::make_shared<CallbackWrapper<void>>(
+    std::bind(&AsyncMotionModel::onStart, this)
+  );
   auto result = callback->getFuture();
-  callback_queue_.addCallback(callback, reinterpret_cast<uint64_t>(this));
+  callback_queue_->addCallback(callback);
   result.wait();
 }
 
 void AsyncMotionModel::stop()
 {
-  if (ros::ok())
+  if (rclcpp::ok())
```

Comment:
This should probably  `context->is_valid()` taken from the owner of the instance, which seems to be the `Optimizer` class downstream. Usually there should only be one `rclcpp::Context` per process as that's the barrier for inter-process communication.

Reply:
https://github.com/locusrobotics/fuse/pull/284/commits/ba64b63714b8acaa5409946b44d84fc7b5ee1d58

Note that the optimizer becomes a child class of rclcpp::Node in the time PR.